### PR TITLE
Easily Selectable History

### PR DESCRIPTION
**Pull request by easong-openai**
This update replaces the previous ratatui history widget with an append-only log so that the terminal can handle text selection and scrolling. It also disables streaming responses, which we'll do our best to bring back in a later PR. It also adds a small summary of token use after the TUI exits.

### CODE REVIEW

File: codex-rs/tui/src/app_event.rs [hunk 1]
```diff
@@ -49,4 +50,9 @@ pub(crate) enum AppEvent {
         query: String,
         matches: Vec<FileMatch>,
     },
+
+    /// Append immutable history lines above the inline viewport. Part of the
+    /// incremental migration to the hybrid scrollback model described in
+    /// `fix-history-plan.md`.
```
Comment (bolinfest) [inline codex-rs/tui/src/app_event.rs]: I think the reference to `fix-history-plan.md` is outdated now?

File: codex-rs/tui/src/chatwidget.rs [hunk 2]
```diff
@@ -53,6 +50,10 @@ pub(crate) struct ChatWidget<'a> {
     token_usage: TokenUsage,
     reasoning_buffer: String,
     answer_buffer: String,
+    // Buffer for streaming assistant answer text; we do not surface partial
```
Comment (bolinfest) [inline codex-rs/tui/src/chatwidget.rs:52]: I feel like this comment should be "attached" to `fn` or `struct`?

File: codex-rs/tui/src/chatwidget.rs [hunk 3]
```diff
@@ -247,50 +257,54 @@ impl ChatWidget<'_> {
...
             EventMsg::AgentMessageDelta(AgentMessageDeltaEvent { delta }) => {
-                if self.answer_buffer.is_empty() {
-                    self.conversation_history
-                        .add_agent_message(&self.config, "".to_string());
-                }
-                self.answer_buffer.push_str(&delta.clone());
-                self.conversation_history
-                    .replace_prev_agent_message(&self.config, self.answer_buffer.clone());
-                self.request_redraw();
+                // Buffer only – do not emit partial lines. This avoids cases
```
Comment (bolinfest) [inline codex-rs/tui/src/chatwidget.rs:280]: Could we keep track of the number of `\n` in `answer_buffer` and add the number introduced by `delta` and flush everything up to the latest `\n`? Maybe in a follow-up PR?

File: codex-rs/tui/src/chatwidget.rs [hunk 4]
```diff
@@ -220,7 +221,10 @@ impl ChatWidget<'_> {
 
         // Only show text portion in conversation history for now.
         if !text.is_empty() {
-            self.conversation_history.add_user_message(text);
+            self.conversation_history.add_user_message(text.clone());
+            if let Some(lines) = self.conversation_history.last_entry_plain_lines() {
+                self.app_event_tx.send(AppEvent::InsertHistory(lines));
+            }
```
Comment (bolinfest) [inline codex-rs/tui/src/chatwidget.rs]: This seems to happen enough that it feels like it should be a helper function.

File: codex-rs/tui/src/history_cell.rs [hunk 5]
```diff
@@ -123,6 +123,30 @@ pub(crate) enum HistoryCell {
...
+            | HistoryCell::BackgroundEvent { view }
+            | HistoryCell::GitDiffOutput { view }
+            | HistoryCell::ErrorEvent { view }
+            | HistoryCell::SessionInfo { view }
+            | HistoryCell::CompletedExecCommand { view }
+            | HistoryCell::CompletedMcpToolCall { view }
+            | HistoryCell::PendingPatch { view }
+            | HistoryCell::ActiveExecCommand { view, .. }
+            | HistoryCell::ActiveMcpToolCall { view, .. } => view.lines.clone(),
+            HistoryCell::CompletedMcpToolCallWithImageOutput { .. } => vec![
```
Comment (bolinfest) [inline codex-rs/tui/src/history_cell.rs:144]: It's not the most high priority thing, but can we not support images anymore?

File: codex-rs/tui/src/insert_history.rs [hunk 6]
```diff
@@ -0,0 +1,183 @@
+use crate::tui; // for the Tui type alias
```
Comment (bolinfest) [inline codex-rs/tui/src/insert_history.rs]: Drop comment?

File: codex-rs/tui/src/lib.rs [hunk 7]
```diff
@@ -47,7 +47,10 @@ mod user_approval_widget;
 
 pub use cli::Cli;
 
-pub fn run_main(cli: Cli, codex_linux_sandbox_exe: Option<PathBuf>) -> std::io::Result<()> {
+pub fn run_main(
+    cli: Cli,
+    codex_linux_sandbox_exe: Option<PathBuf>,
+) -> std::io::Result<codex_core::protocol::TokenUsage> {
```
Comment (bolinfest) [inline codex-rs/tui/src/lib.rs:53]: I think it might be helpful to introduce a struct like `FinalOutput` that has `TokenUsage` (and maybe `final_message: Option<String>`) as a field.

Then I would have it `impl Display for FinalOutput` with the implementation that corresponds to this code:

```
            println!(
                "Token usage: total={} input={}{} output={}{}",
                usage.total_tokens,
                usage.input_tokens,
                usage
                    .cached_input_tokens
                    .map(|c| format!(" (cached {c})"))
                    .unwrap_or_default(),
                usage.output_tokens,
                usage
                    .reasoning_output_tokens
                    .map(|r| format!(" (reasoning {r})"))
                    .unwrap_or_default()
            );
```

so that you can just do:

```
println!("{final_output}");
```

rather than copy/paste the two implementations.

File: codex-rs/tui/src/slash_command.rs [hunk 8]
```diff
@@ -15,17 +15,13 @@ pub enum SlashCommand {
...
     Quit,
-    ToggleMouseMode,
 }
 
 impl SlashCommand {
     /// User-visible description shown in the popup.
     pub fn description(self) -> &'static str {
         match self {
             SlashCommand::New => "Start a new chat.",
-            SlashCommand::ToggleMouseMode => {
```
Comment (bolinfest) [inline codex-rs/tui/src/slash_command.rs:26]: Nice! We can also remove `disable_mouse_capture` in the docs (`config.md`) and `Config` code.

File: codex-rs/tui/src/insert_history.rs [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+use unicode_width::UnicodeWidthChar;
+
+/// Insert a batch of history lines into the terminal scrollback above the
+/// inline viewport.
+///
+/// The incoming `lines` are the logical lines supplied by the
+/// `ConversationHistory`. They may contain embedded newlines and arbitrary
+/// runs of whitespace inside individual [`Span`]s. All of that must be
+/// normalised before writing to the backing terminal buffer because the
+/// ratatui [`Paragraph`] widget does not perform soft‑wrapping when used in
```
Comment (bolinfest) [inline codex-rs/tui/src/insert_history.rs:17]: So even if we explicitly use the `wrap()` method of `Paragraph`, there is nothing we can do?

File: codex-rs/tui/src/insert_history.rs [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+                *width = 0;
+            };
+
+        // Iterate spans tokenising into words and whitespace so wrapping can
+        // happen at word boundaries.
+        for span in logical.spans.into_iter() {
+            let style = span.style;
+            let mut buf_word = String::new();
+            let mut buf_space = String::new();
+            let flush_word = |word: &mut String,
```
Comment (bolinfest) [inline codex-rs/tui/src/insert_history.rs]: Can we pull out the appropriate arguments so this can be a top-level function instead of a closure? I think that would make things a bit easier to follow.

Comment (bolinfest) [reply codex-rs/tui/src/insert_history.rs]: Alternatively, maybe it would be easier to create a struct with some methods on it to manage this entire computation? Then things can be read off `self` instead of passed in?

**Review by bolinfest — CHANGES_REQUESTED** [review]
Overall, this is really great!

Unsurprisingly, the implementation of `insert_history_lines()` is what gives me the most pause. I haven't played with these APIs/edge cases myself, so I assume all the complexity is necessary to make things work.

It would certainly be helpful to add unit tests in the near future to be sure that we are catching all these edge cases.

File: codex-rs/tui/src/chatwidget.rs (hunk 3 above)
Comment (easong-openai) [reply codex-rs/tui/src/chatwidget.rs:280]: Yes - I think this will be a followup PR though. We might be able to do some fancy stuff to get chunk streaming and format as we go.

File: codex-rs/tui/src/history_cell.rs (hunk 5 above)
Comment (easong-openai) [reply codex-rs/tui/src/history_cell.rs:144]: We will support images, but not immediately (which I think is the same as the current UX?)

File: codex-rs/tui/src/insert_history.rs (hunk 9 above)
Comment (easong-openai) [reply codex-rs/tui/src/insert_history.rs:17]: I intend to refactor this code significantly to try to support live updates before committing a final version (which would allow us to bring back streaming/command outputs).

File: codex-rs/tui/src/chatwidget.rs (hunk 4 above)
Comment (easong-openai) [reply codex-rs/tui/src/chatwidget.rs]: 👍

File: codex-rs/tui/src/slash_command.rs (hunk 8 above)
Comment (easong-openai) [reply codex-rs/tui/src/slash_command.rs:26]: 👍

File: codex-rs/tui/src/insert_history.rs (hunk 10 above)
Comment (easong-openai) [reply codex-rs/tui/src/insert_history.rs]: Struct makes sense to me!

File: codex-rs/core/src/protocol.rs [hunk 11]
```diff
@@ -4,9 +4,10 @@
 //! between user and agent.
 
 use std::collections::HashMap;
+use std::fmt;
 use std::path::Path;
 use std::path::PathBuf;
-use std::str::FromStr;
+use std::str::FromStr; // Added for FinalOutput Display implementation
```
Comment (bolinfest) [inline codex-rs/core/src/protocol.rs:10]: Remove comment?

File: codex-rs/core/src/protocol.rs [hunk 12]
```diff
@@ -355,6 +356,36 @@ pub struct TokenUsage {
     pub total_tokens: u64,
 }
 
+#[derive(Debug, Clone, Deserialize, Serialize)]
```
Comment (bolinfest) [inline codex-rs/core/src/protocol.rs:359]: protocol.rs should only have things that are part of the protocol between the business logic and the UI layer. This seems more appropriate for codex-rs/common?

File: codex-rs/tui/src/chatwidget.rs (hunk 2 above)
Comment (bolinfest) [reply codex-rs/tui/src/chatwidget.rs:52]: Note doc comments in Rust should have `///`.

**Review by bolinfest — APPROVED** [review]
Just a few small suggestions: great work!
